src/goto-symex/shadow_memory_util.cpp

```diff
@@ -183,3 +183,311 @@ void log_cond(
     });
 #endif
 }
+
+void replace_invalid_object_by_null(exprt &expr)
+{
+  if(
+    expr.id() == ID_symbol && expr.type().id() == ID_pointer &&
+    (id2string(to_symbol_expr(expr).get_identifier()).rfind("invalid_object") !=
+       std::string::npos ||
+     id2string(to_symbol_expr(expr).get_identifier()).rfind("$object") !=
+       std::string::npos))
+  {
+    expr = null_pointer_exprt(to_pointer_type(expr.type()));
+    return;
+  }
+  Forall_operands(it, expr)
+  {
+    replace_invalid_object_by_null(*it);
+  }
+}
+
+const exprt &
+get_field_init_expr(const irep_idt &field_name, const goto_symex_statet &state)
+{
+  auto field_type_it = state.shadow_memory.fields.local_fields.find(field_name);
+  if(field_type_it != state.shadow_memory.fields.local_fields.end())
+  {
+    return field_type_it->second;
+  }
+  field_type_it = state.shadow_memory.fields.global_fields.find(field_name);
+  CHECK_RETURN(field_type_it != state.shadow_memory.fields.global_fields.end());
+  return field_type_it->second;
+}
+
+/// Given a pointer expression `address` checks if any expr in value_set is
+/// compatible
+bool contains_compatible_address(
```

Comment:
This is misnamed and the comment above is wrong. This should be called `contains_null_or_invalid`.
Also, I think there's potential for simplifications as together with the invalid-by-null replacement the unknown case can probably go away (in a future PR).